errors.go

```diff
@@ -100,6 +105,21 @@ func (p *Error) Error() string {
 	return output.String()
 }
 
+// ErrorChain returns an extended error message which includes a complete causal chain.
+// When errors are marshalled certain information is lost (e.g. the 'cause').  This means if an error travels through
+// a number of services (and it's potentially augmented at each hop) that the core error message may be lost.
+func (p *Error) ErrorChain() string {
```

Comment:
It might also be convenient to add a `func ErrorChain(err error) string`, so that users can call it on an `error` without needing to explicitly type asset to a terrors.Error.

Reply:
Good idea, I'll add this.

Reply:
Actually, I'm having second thoughts about this.

Even though the idea of dropping info relating to low level errors hasn't been well received, I do think there was sound reasoning behind it.  If we make it too easy for people to call `ErrorChain(...)` then there's a risk that people default to this, and this then increases the risk that high level code becomes coupled to errors which are generated by low-level/indirect dependencies.

My intention with this PR is really just so that we have the ability to log the causal chain.  I don't think people should be interacting with this in any other way.

Reply:
Fair enough. Let's leave it for now - it's something that could be added in the future if there's a lot of demand for it.